Refactor string_constraint_generatort

romainbrenguier opened the pull request:
The main goal is to extract all the add_axioms functions from the `string_constraint_generatort` class so that they can be reused from `string_builtin_functiont` classes. This has also the benefit of making the `string_constraint_generatort` class much smaller, and what is modified by these function is now clearer through their interface.
(warning: if you review commit by commit, the one before last is quite big)

romainbrenguier committed:
Extract witness map from string_constraint_generator

romainbrenguier committed:
Remove unused get_boolean/index_symbols

romainbrenguier committed:
Get rid of fresh_exist/univ_symbol

These are now equivalent to fresh_symbol

romainbrenguier committed:
Use from_integer instead of constant_char

constant_char was just an alias for from_integer so there is no
advantage in keeping it.

romainbrenguier committed:
Extract is_positive from generator class

romainbrenguier committed:
Make int_of_hex_char static

romainbrenguier committed:
Make is_high_surrogate and is_low_surrogate static

romainbrenguier committed:
Make char_equals_ignore_case, to_integer_or_default static

romainbrenguier committed:
Extract find and of_argument of array_pool class

These are not using private field of the class but only calling public
functions.

romainbrenguier committed:
Remove unused find function

romainbrenguier committed:
Move created_strings field to array_pool

Since this is only updated when we do array_pool operations it makes
sense to move it there.

romainbrenguier committed:
Call array_pool.created_strings() directly

instead of generator.get_created_strings which does the same thing.

romainbrenguier committed:
Remove use of fresh_boolean

This is just an alias for the default behavior of fresh_symbol

romainbrenguier committed:
Extract a string_constraintst struct from generator

This better seperates the object we are constructing from the tools we
are using to construct it and will make it easier to make constraint
generation functions static.

romainbrenguier committed:
Add a merge function for string_constraints

romainbrenguier committed:
Make constraint_generation_function return constraints

This make the role of these functions clearer.

romainbrenguier committed:
Add fresh_symbol argument to add_axioms functions

This will be necessary to make the functions static

romainbrenguier committed:
Make add_axioms functions static

This makes how the functions interact with the generator clearer, and
make it easier to reuse the functions outside of the class and to unit
test them.

romainbrenguier committed:
Make char_array_of_pointer static

romainbrenguier committed:
Make get_string_expr static

romainbrenguier committed:
Move method definition to cpp file

romainbrenguier committed:
Remove unused functions

romainbrenguier committed:
Adapt unit tests for constraint generator changes

romainbrenguier committed:
Add missing doxygen argument documentation

martin-cs commented on src/solvers/refinement/string_constraint_generator.h:
```diff
@@ -115,10 +115,6 @@ class string_constraint_generatort final
-  symbol_exprt fresh_univ_index(const irep_idt &prefix, const typet &type);
-
-  symbol_exprt fresh_exist_index(const irep_idt &prefix, const typet &type);
-
```
From a formal point of view, having these being handled the same is a little concerning.  I'm guessing this is just a naming artefact rather than something deeper.

martin-cs reviewed (approved):
This PR seems to do what it describes.  I don't have a clear enough picture of the design to say whether this is the right thing to do or not.

romainbrenguier committed:
Remove constraint_generator qualifier in README

These methods are now non-member functions.

tautschnig reviewed (approved):
Approving of the changes to the README file. For everything else I have to trust that it's right (CI says it currently isn't).

romainbrenguier committed:
Adapt debugging function for cleaned-up generator

Some functions that were used for debugging have been remove from string
constraint generator.
We adapt to that by adding a vector tracking created symbols when the
DEBUG flag is active.

allredj commented on src/solvers/refinement/string_constraint_generator.h:
```diff
@@ -116,13 +116,6 @@ class string_constraint_generatort final
   /// Set of strings that have been created by the generator
   const std::set<array_string_exprt> &get_created_strings() const;
 
-  exprt get_witness_of(
```
Not sure extracting the witnesses is an improvement. What is the rationale behind that?

romainbrenguier replied:
Conceptually it's a part of the solver procedure, and it's not part of the constraints. I want to make the distinction clear between what we are solving (string constraints) and how (string refinement).

allredj commented on src/solvers/refinement/string_constraint_generator_comparison.cpp:
```diff
@@ -78,30 +78,34 @@ exprt string_constraint_generatort::add_axioms_for_equals(
+  exprt char2,
+  exprt char_a,
+  exprt char_A,
+  exprt char_Z)
```
These should all be const refs.

allredj commented on src/solvers/refinement/string_constraint_generator.h:
```diff
@@ -84,6 +76,15 @@ class array_poolt final
 };
 
+array_string_exprt
+find(array_poolt &array_pool, const refined_string_exprt &str);
```
`array_pool` should probably be const.

romainbrenguier replied:
A new entry is created if `str` was not already present, so the pool cannot be const

allredj commented on src/solvers/refinement/string_constraint_generator.h:
```diff
@@ -90,6 +90,18 @@ class array_poolt final
 
+/// Collection of constraints of different types: existential formulas,
+/// universal formulas, and "not contains" (universal with one alternation).
+struct string_constraintst final
```
The name choice is not great as it looks very much like `string_constraintt` and makes it fairly confusing.

romainbrenguier replied:
agreed, any suggestion?

romainbrenguier replied:
`string_constraint_sett`?

allredj replied:
How painful would it be to keep this `string_constraintst` and rename `string_constraintt` to something else?

romainbrenguier replied:
Not very, but if we change we should try to make it a definitive choice.
I guess any choice in the regexp: (`universal_` | `quantified_`)? (`string` | `array`) `_` (`property` | `formula` | `constraint`) `t` could work

allredj replied:
`string_constraintt` -> `universal_string_constraintt` would make sense to me.

allredj commented on src/solvers/refinement/string_builtin_function.h:
```diff
@@ -128,9 +128,12 @@ class string_concat_char_builtin_functiont
-    return generator.add_axioms_for_concat_char(result, input, character);
+    auto pair = generator.add_axioms_for_concat_char(result, input, character);
+    pair.second.existential.push_back(equal_exprt(pair.first, return_code));
+    return pair.second;
```
I don't understand what we're doing here, and that commit introduces a lot of code duplication. What exactly are we doing?

romainbrenguier replied:
Yes this is a bit boilerplate, I will try to refactor that

romainbrenguier replied:
I've added commits at the end (the "Move add_axioms_*" ones) which show how the `constraints()` codes should look like. I haven't done it for concat and insert yet because it requires getting rid of the other usages of these add_axioms functions first.

allredj reviewed (approved):
Looks good. Please see my question about having the string constraint functions returning the constraints. I don't really understand why and how we do that.

romainbrenguier committed:
Move add_axioms_for_concat_char to builtin functions

The constraints are now added directly by the string_builtin_function
objects, so the code can be moved there.

romainbrenguier committed:
Move add_axioms_for_set_char to builtin_functions

romainbrenguier committed:
Move add_axioms_for_to_lower/upper_case to builtin functions

romainbrenguier committed:
Add string_to_upper_case_builtin_functiont constructor

romainbrenguier committed:
Clean-up doxygen errors

allredj reviewed (commented):
This PR failed Diffblue compatibility checks (cbmc commit: cddfd184a5629cf6240ca4457df009e946a6038f).
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

allredj reviewed (commented):
Passed Diffblue compatibility checks (cbmc commit: a6fe4e1f3bf7b8624142a7b4b435871d420799c7).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/84271237
